### ISSUE TITLE
Operating etcd clusters for Kubernetes

### ISSUE THREAD

**Issue by FerminCastro** [kubernetes/website#31822]
See etcd issue https://github.com/etcd-io/etcd/issues/13724 this is preventing us from upgrading a kubeadm control plane from 1.21.9 to 1.22. We need confirmation that we can just update the etcd image to 3.5.2 or a detailed alternative to upgrae the internal etcd cluster version BEFORE upgrading kubeadm/the control plane....

**Comment by neolit123** [comment]
there is no etcd 3.5.2 in the k8s GCR bucket, yet.
https://console.cloud.google.com/gcr/images/k8s-artifacts-prod/us/etcd

you can log an issue for that in kubernetes/kubernetes.
there was PR but it got closed:
https://github.com/kubernetes/kubernetes/pull/107917

to use a custom etcd images in kubeadm you can edit the kube-system/kubead-config config map that holds the ClusterConfiguration:

```
apiVersion: kubeadm.k8s.io/v1beta3
kind: ClusterConfiguration
...
etcd:
  local:
    imageRepository: "your.registry.com"
    imageTag: "3.5.2"
```

then modify the /etc/kubernetes/manifests/etcd.yaml on CP machines one at a time.

https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/#custom-images

for more support questions please use:
https://github.com/kubernetes/kubernetes/blob/master/SUPPORT.md

/kind support
/close
/sig cluster-lifecycle api-machinery

**Comment by FerminCastro** [comment]
Thanks, that was useful. I think that since there have been multiple occasions and issues where there is a lag between etcd bugs and k8 adopting the etcd version with the fix, an "upgrade etcd" section would make sense in the "Operating etcd clusters for Kubernetes" doc, would you agree?

**Comment by neolit123** [comment]
> We need confirmation that we can just update the etcd image to 3.5.2 or a detailed alternative to upgrae the internal etcd cluster version BEFORE upgrading kubeadm/the control plane....

we don't have 3.5.2 tests yet, so technically we cannot give you *any* confirmation that 3.5.2 works.

> Thanks, that was useful. I think that since there have been multiple occasions and issues where there is a lag between etcd bugs and k8 adopting the etcd version with the fix, an "upgrade etcd" section would make sense in the "Operating etcd clusters for Kubernetes" doc, would you agree?

that page is k8s + etcd generic and not about kubeadm.
what we have for kubeadm custom images is here:
https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/#custom-images

changing the etcd image on a running cluster is not something that we recommend.
advanced users follow the kubeadm implementations details:
https://kubernetes.io/docs/reference/setup-tools/kubeadm/implementation-details/
and can change the image, but they are in unsupported territory after that.

**Comment by neolit123** [comment]
if you'd like to suggest changes for https://kubernetes.io/docs/tasks/administer-cluster/configure-upgrade-etcd/ that don't include kubeadm, you can re-open this issue though.

**Comment by FerminCastro** [comment]
We really don't care where this is described ... since there is no documentation specific to etcd created by kubeadm we thought the nearest possible would be the generic k8 admin etcd one...seems like we are in an endless loop here and being pointed to a different area in each area jump

SUMMARY:
The internal etcd cluster created by kubeadm has a bug. K8 (not just the control plane but the entire K8 cluster) cannot be upgraded because of that etcd issue. And there is no formal support nor document to upgrade etcd to that version for the K8 cluster... so basically ALL FIXES in K8 1.22 and onwards cannot be consumed by users like us until K8 supports etcd 3.5.2 ... 

If newer/custom etcd images are not formally supported/recommended in the control place created by kubeadm I don't think it will help much to add the steps there to upgrade it... Not a good model of dependencies. We will ask for PR again

Thanks

**Comment by FerminCastro** [comment]
OK, again since there is no specific kubeadm etcd upgrade documentation, let me clarify here HOW we finally were able to get passed this. The steps are clearly far from "clean" and clearly call for a formal procedure to update etcd when a critical etcd bug arises in a kubeadm cluster. Specially if that etcd bug precludes k8 upgrade (which basically precludes any k8 fixes in your system) 

Following the indications above, we manually updated the etcd.yaml and kubeadm-config cm to use the 3.5.2 image
[opc@olk8-m1 ~]$ k get cm kubeadm-config -n kube-system -o yaml | | grep etcd: -A 3
    etcd:
      local:
        imageRepository: gcr.io/etcd-development
        imageTag: v3.5.2

In all Control Plane nodes:

[opc@olk8-m1 ~]$ sudo grep image: /etc/kubernetes/manifests/etcd.yaml
       image: gcr.io/etcd-development/etcd:v3.5.2

We restarted all control plane pods by moving manifests back and forth. After this, we confirmed that zombie etcd memebrs as reported in https://github.com/etcd-io/etcd/issues/13724 were gone. However we could not upgrade the control plane from 1.21 to 1.22 with this image setting for etcd. kubeadm upgrade would time out trying to get the correct status from etcd pods (we have the logs for this error if needed). So we basically had to revert the etcd image HOPING that the clean up of zombie etcd members would remain in the physical etcd store (and crossing fingers that any alterations made by etcd 3.5.2 would not break when going back to 3.5.0). The result was that kubeadm upgrade proceeded once back to etcd 3.5.0 except in the last control plane in which the same failure about zombie members came back. Since upgrade had completed successfully in the first 2 nodes, we just changed back again the etcd image to 3.5.2 and restarted the entire control plane again. After that, the upgrade proceeded properly in the last control plane and we got everything on k8 1.22.. At this point and to be in "supported mode" we only had to revert the image back to 3.5.0

[opc@olk8-m1 ~]$ sudo grep image: /etc/kubernetes/manifests/etcd.yaml
    image: k8s.gcr.io/etcd:3.5.0-0

It is possible that we had broken something in between that made all this such a nightmare, but still  it looks clear that there needs to be a faster and well defined procedure to catch up with etcd fixes in older k8 systems (rather than waiting for k8 to catch up and perform a ton of k8 upgrades to get there) Specially if etcd bugs BREAK k8 upgrade :-)

Thanks again for the help
